[java] UseStandardCharsets: ArrayIndexOutOfBoundsException in line 81

<!-- Please, prefix the report title with the language it applies to within brackets, such as [java] or [apex].
If not specific to a language, you can use [core]. -->

**Affects PMD Version: 7.25.0** 

**Description:**

When Running PMD on the following source code (snippet of the actual file), then PMD throws an exception. This worked in previous PMD releases.

**Exception Stacktrace:**

```
Exception applying rule UseStandardCharsets on file /home/runner/work/codingstyle/codingstyle/src/test/java/edu/hm/hafner/util/ValidationUtilitiesTest.java, continuing with next rule
org.apache.commons.lang3.exception.ContextedRuntimeException: java.lang.IndexOutOfBoundsException: Index 1 out of bounds for length 1
Exception Context:
	[1:Rule applied on node=!debug only! [MethodCall:11:9]assertThat(allCharsets).isNotEmpty().contains("UTF-8", "ISO-8859-1")]
---------------------------------
    at net.sourceforge.pmd.util.AssertionUtil.contexted (AssertionUtil.java:256)
    at net.sourceforge.pmd.lang.rule.internal.RuleApplicator.applyOnIndex (RuleApplicator.java:79)
    at net.sourceforge.pmd.lang.rule.internal.RuleApplicator.apply (RuleApplicator.java:57)
    at net.sourceforge.pmd.lang.rule.internal.RuleSets.apply (RuleSets.java:183)
    at net.sourceforge.pmd.lang.impl.PmdRunnable.processSource (PmdRunnable.java:140)
    at net.sourceforge.pmd.lang.impl.PmdRunnable.run (PmdRunnable.java:80)
    at java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:545)
    at java.util.concurrent.FutureTask.run (FutureTask.java:328)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1090)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:614)
    at java.lang.Thread.run (Thread.java:1474)
Caused by: java.lang.IndexOutOfBoundsException: Index 1 out of bounds for length 1
    at jdk.internal.util.Preconditions.outOfBounds (Preconditions.java:100)
    at jdk.internal.util.Preconditions.outOfBoundsCheckIndex (Preconditions.java:106)
    at jdk.internal.util.Preconditions.checkIndex (Preconditions.java:302)
    at java.util.Objects.checkIndex (Objects.java:365)
    at java.util.ArrayList.get (ArrayList.java:428)
    at net.sourceforge.pmd.lang.java.rule.bestpractices.UseStandardCharsetsRule.lambda$checkIthArgument$2 (UseStandardCharsetsRule.java:81)
    at java.util.stream.ReferencePipeline$2$1.accept (ReferencePipeline.java:196)
    at java.util.stream.ReferencePipeline$2$1.accept (ReferencePipeline.java:197)
    at java.util.stream.ReferencePipeline$2$1.accept (ReferencePipeline.java:197)
    at java.util.stream.ForEachOps$ForEachOp$OfRef.accept (ForEachOps.java:186)
    at java.util.stream.ReferencePipeline$3$1.accept (ReferencePipeline.java:214)
    at java.util.stream.ReferencePipeline$2$1.accept (ReferencePipeline.java:197)
    at java.util.ArrayList$ArrayListSpliterator.forEachRemaining (ArrayList.java:1716)
    at java.util.stream.AbstractPipeline.copyInto (AbstractPipeline.java:570)
    at java.util.stream.AbstractPipeline.wrapAndCopyInto (AbstractPipeline.java:560)
    at java.util.stream.ForEachOps$ForEachOp.evaluateSequential (ForEachOps.java:153)
    at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential (ForEachOps.java:176)
    at java.util.stream.AbstractPipeline.evaluate (AbstractPipeline.java:265)
    at java.util.stream.ReferencePipeline.forEach (ReferencePipeline.java:632)
    at java.util.stream.ReferencePipeline$7$1FlatMap.accept (ReferencePipeline.java:293)
    at java.util.Iterator.forEachRemaining (Iterator.java:133)
    at java.util.Spliterators$IteratorSpliterator.forEachRemaining (Spliterators.java:1939)
    at java.util.stream.AbstractPipeline.copyInto (AbstractPipeline.java:570)
    at java.util.stream.AbstractPipeline.wrapAndCopyInto (AbstractPipeline.java:560)
    at java.util.stream.ReduceOps$ReduceOp.evaluateSequential (ReduceOps.java:921)
    at java.util.stream.ReduceOps$5.evaluateSequential (ReduceOps.java:258)
    at java.util.stream.ReduceOps$5.evaluateSequential (ReduceOps.java:248)
    at java.util.stream.AbstractPipeline.evaluate (AbstractPipeline.java:265)
    at java.util.stream.ReferencePipeline.count (ReferencePipeline.java:750)
    at net.sourceforge.pmd.lang.java.rule.bestpractices.UseStandardCharsetsRule.checkIthArgument (UseStandardCharsetsRule.java:82)
    at net.sourceforge.pmd.lang.java.rule.bestpractices.UseStandardCharsetsRule.visit (UseStandardCharsetsRule.java:54)
    at net.sourceforge.pmd.lang.java.rule.bestpractices.UseStandardCharsetsRule.visit (UseStandardCharsetsRule.java:31)
    at net.sourceforge.pmd.lang.java.ast.ASTMethodCall.acceptVisitor (ASTMethodCall.java:69)
    at net.sourceforge.pmd.lang.java.ast.AbstractJavaNode.acceptVisitor (AbstractJavaNode.java:38)
    at net.sourceforge.pmd.lang.java.rule.AbstractJavaRule.apply (AbstractJavaRule.java:30)
    at net.sourceforge.pmd.lang.rule.RuleReference.apply (RuleReference.java:413)
    at net.sourceforge.pmd.lang.rule.internal.RuleApplicator.applyOnIndex (RuleApplicator.java:77)
    at net.sourceforge.pmd.lang.rule.internal.RuleApplicator.apply (RuleApplicator.java:57)
    at net.sourceforge.pmd.lang.rule.internal.RuleSets.apply (RuleSets.java:183)
    at net.sourceforge.pmd.lang.impl.PmdRunnable.processSource (PmdRunnable.java:140)
    at net.sourceforge.pmd.lang.impl.PmdRunnable.run (PmdRunnable.java:80)
    at java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:545)
    at java.util.concurrent.FutureTask.run (FutureTask.java:328)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1090)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:614)
    at java.lang.Thread.run (Thread.java:1474)
```

**Code Sample demonstrating the issue:**

```java
package edu.hm.hafner.util;

import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.*;

class ValidationUtilitiesTest {
    @Test
    void shouldContainDefaultCharsets() {
        var allCharsets = getAllCharsets();
        assertThat(allCharsets).isNotEmpty().contains("UTF-8", "ISO-8859-1");
    }

    private String getAllCharsets() {
        return "UTF-8";
    }
}
```

**Steps to reproduce:**

See https://github.com/uhafner/codingstyle/pull/1905

**Running PMD through:** *Maven*

<!-- If relevant, also include your JDK and OS information, e.g. for ClassNotFoundException, LinkageError, reflection failures, etc. -->

--- a/docs/pages/release_notes.md
+++ b/docs/pages/release_notes.md
@@ -42,6 +42,7 @@ This is a {{ site.pmd.release_type }} release.
   * [#6641](https://github.com/pmd/pmd/issues/6641): \[cpp]: IndexOutOfBoundsException in CPD when a duplication is at end of file with UTF8-BOM
 * java-bestpractices
   * [#6692](https://github.com/pmd/pmd/issues/6692): \[java] ForLoopCanBeForeach: inconsistent detection between i += 1 and i = i + 1 update forms
+  * [#6782](https://github.com/pmd/pmd/issues/6782): \[java] UseStandardCharsets: ArrayIndexOutOfBoundsException in line 81
 * java-codestyle
   * [#6239](https://github.com/pmd/pmd/issues/6239): \[java] UseDiamondOperator: False positive with Guice TypeLiteral
   * [#6775](https://github.com/pmd/pmd/issues/6775): \[java] UselessParentheses: False negative when on the right-hand side of an assignment statement

--- a/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/UseStandardCharsetsRule.java
+++ b/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/UseStandardCharsetsRule.java
@@ -69,6 +69,10 @@ public class UseStandardCharsetsRule extends AbstractJavaRulechainRule {
     }
 
     private void checkIthArgument(RuleContext ctx, InvocationNode call, int index) {
+        if (index >= call.getMethodType().getArity()) {  // this can happen if we're looking at varargs!
+            return;
+        }
+
         Object callArgument = call.getArguments().get(index).getConstValue();
         if (callArgument instanceof String) {
             String stringArgument = (String) callArgument;

--- a/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/UseStandardCharsetsRule.java
+++ b/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/UseStandardCharsetsRule.java
@@ -69,25 +69,28 @@ public class UseStandardCharsetsRule extends AbstractJavaRulechainRule {
     }
 
     private void checkIthArgument(RuleContext ctx, InvocationNode call, int index) {
-        if (index >= call.getMethodType().getArity()) {  // this can happen if we're looking at varargs!
+        JMethodSig callSignature = call.getMethodType();
+        if (index >= callSignature.getArity()) {  // this can happen if we're looking at varargs!
             return;
         }
 
         Object callArgument = call.getArguments().get(index).getConstValue();
-        if (callArgument instanceof String) {
-            String stringArgument = (String) callArgument;
-            if (standardCharsetExists(call.getLanguageVersion(), stringArgument)) {
-                JMethodSig callSignature = call.getMethodType();
-                Stream<JMethodSig> otherSignatures = streamMethodSignatures(call);
-                long count = otherSignatures
-                        .filter(sig -> Modifier.isPublic(sig.getModifiers()))
-                        .filter(sig -> isSameSignatureExcept(callSignature, sig, index))
-                        .filter(sig -> CHARSET.equals(sig.getFormalParameters().get(index).toString()))
-                        .count();
-                if (count > 0) {
-                    ctx.addViolation(call);
-                }
-            }
+        if (!(callArgument instanceof String)) {
+            return;
+        }
+        String stringArgument = (String) callArgument;
+        if (!standardCharsetExists(call.getLanguageVersion(), stringArgument)) {
+            return;
+        }
+
+        Stream<JMethodSig> otherSignatures = streamMethodSignatures(call);
+        long count = otherSignatures
+                .filter(sig -> Modifier.isPublic(sig.getModifiers()))
+                .filter(sig -> isSameSignatureExcept(callSignature, sig, index))
+                .filter(sig -> CHARSET.equals(sig.getFormalParameters().get(index).toString()))
+                .count();
+        if (count > 0) {
+            ctx.addViolation(call);
         }
     }
 

--- a/pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/UseStandardCharsets.xml
+++ b/pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/UseStandardCharsets.xml
@@ -457,4 +457,42 @@ public class Foo {
         ]]></code>
         <source-type>java 21</source-type>
     </test-code>
+
+    <test-code>
+        <description>varargs</description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+public class Foo {
+    public void foo() {
+        bar("UTF-8", "ISO-8859-1");
+    }
+
+    public void bar(String... x) { }
+}
+        ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>[java] ArrayIndexOutOfBoundsException in rule UseStandardCharsetsRule.java:81 #6782</description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+package edu.hm.hafner.util;
+
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.Assertions.*;
+
+class ValidationUtilitiesTest {
+    @Test
+    void shouldContainDefaultCharsets() {
+        var allCharsets = getAllCharsets();
+        assertThat(allCharsets).isNotEmpty().contains("UTF-8", "ISO-8859-1");
+    }
+
+    private String getAllCharsets() {
+        return "UTF-8";
+    }
+}
+        ]]></code>
+    </test-code>
 </test-data>

